esx_service/vmdk_ops.py

```diff
@@ -777,31 +776,90 @@ def disk_attach(vmdk_path, vm):
+            controller_key = pvsci[idx].key
+            taken = set([dev.unitNumber
                      for dev in devices
                      if type(dev) == vim.VirtualDisk and dev.controllerKey ==
                      controller_key])
-        # search in 15 slots, with unit_number 7 reserved for scsi controller
-        availSlots = (set(range(0, 7)) | set(range(8, 16))) - taken
-
-        if len(availSlots) == 0:
-            msg = "Failed to place new disk - out of disk slots"
-            logging.error(msg + " VM=%s", vm.config.uuid)
-            return err(msg)
-
-        disk_slot = availSlots.pop()
-        logging.debug("controller_key = %d slot = %d", controller_key, disk_slot)
+            # search in 15 slots, with unit_number 7 reserved for scsi controller
+            availSlots = (set(range(0, 7)) | set(range(8, 16))) - taken
+            logging.debug("idx=%d controller_key=%d availSlots=%d", idx, controller_key, len(availSlots))
+
+            if len(availSlots) == 0:
+                idx = idx+1
+            else:     
+                disk_slot = availSlots.pop()
+                pci_slot_number = get_controller_pci_slot(vm, pvsci[idx],
+                                                          offset_from_bus_number)
+                 
+                logging.debug("Find an available slot: controller_key = %d slot = %d", controller_key, disk_slot)
+
+        # search through all the existing pvscsi controller, and cannot find a slot
+        # try to add another pvscsi controller if possible        
+        if ((disk_slot is None) and (len(availSlots) == 0)):
+            disk_slot = 0  # starting on a fresh controller
+            if len(controllers) >= max_scsi_controllers:
+                msg = "Failed to place new disk - out of disk slots"
+                logging.error(msg + " VM=%s", vm.config.uuid)
+                return err(msg)
+                    
+            # find empty bus slot for the controller:
+            taken = set([c.busNumber for c in controllers])
+            avail = set(range(0, max_scsi_controllers)) - taken
+
+            key = avail.pop()  # bus slot
+            controller_key = key + offset_from_bus_number
+            controller_spec = vim.VirtualDeviceConfigSpec(
+            operation='add',
+            device=vim.ParaVirtualSCSIController(key=controller_key,
+                                                 busNumber=key,
+                                                 sharedBus='noSharing', ), )
+            # changes spec content goes here
+            pvscsi_change = []
+            pvscsi_change.append(controller_spec)
+            spec = vim.vm.ConfigSpec()
+            spec.deviceChange = pvscsi_change
+
+            try:
+                wait_for_tasks(si, [vm.ReconfigVM_Task(spec=spec)])
+            except vim.fault.VimFault as ex:
+                msg=("Failed to add PVSCSI Controller: %s", ex.msg)
+                return err(msg)
+            
```

Comment:
Please put this VM reconfig code to add a new pvscsi controller into a separate function like add_scsi_controller() which defaults to always adding pvscsi controllers. Called only if disk_slot is None.